docs/Configuration.rst
==========================================
Configuration
==========================================
**to be updated, mostly accurate**

The configuration file enables the user to adapt the simulation to ones need, as listed below. The input file should, however, not be overlooked, as it summarizes all simulation input data and especially enables a sensitivity analysis of a number of parameters. The Input file as well as the sensitivity parameters will be discussed in separate entry.

Coding mode
-----------
To simplify the simulation during coding (speeding up the calculations), **coding_process** can be set to True, shortening the analyzed time frame (_time_start_, _time_end_, _frequency_) and disabling white noise (not coded jet neither for irradiation nor for demand, set in input file with _white_noise_demand_, _white_noise_weather_).::

        coding_process              = True

Restoring results
-----------------
It is possible to resume a sensitivity analysis at the point of interruption by checking::

        restore_oemof_if_existant   = True

Otherwise, all previous results will be deleted and should be manually moved to another folder, if they are to be stored long-term!

For this setting to work, **setting_save_oemofresults** also has to be set to True.

Please keep in mind that the oemof-files are saved with the name of their case and, if present, their sensitivity case. Changing values in the config-file or the input values will therefore will not generate a new oemof-file! When changing settings or values that are not sensitivity values, _restore_oemof_if_existant_ should be disabled to result in a new simulation!.

Simulated cases
---------------
* Base case OEM is by default performed without minimal loading of generators, enabling their sizing. If a minimal loading has to be taken into account, then setting  **base_case_with_min_loading** fixes the generator capacity to the demand peak value (without security margin).::

            base_case_with_min_loading = True


* Checking all operational modes (cases) to be analyzed in **simulated_cases**. A base case OEM to determine necessary capacities is always performed.::

        simulated_cases = {
            'offgrid_fixed': True,            # dispatch with base oem capacities
            'interconnected_buy': True,       # dispatch with interconnected main grid (only consumption)
            'interconnected_buysell': True,   # dispatch with interconnected main grid (consumption+feedin)
            'oem_grid_tied_mg': True,         # optimization of grid-tied mg (consumption+feedin)
            'buyoff': False,                  # NOT IMPLEMENTED
            'parallel': False,                # NOT IMPLEMENTED
            'adapted': False,                 # NOT IMPLEMENTED
        }

Additional settings
___________________
* A sensitivity constraint is added with:::

        include_stability_constraint = True

* Enabling / disabling supply shortage at the project site with **allow_shortage**, especially important when including noise. The maximal share of demand not supplied (shortage) and variable costs can be defined in input_values.py::

        allow_shortage              = False

* If **setting_batch_capacity** is True, then the results of the OEM will be processed in regard to batch capacity sizes. The batch capacities for system components (PV, storage, genset) are defined with _round_to_batch_. The values are always round up to the next batch size (also meaning that an OEM result of 15 kW will be 15kW+batch capacity). If batch capacities are not activated, the dispatch optimization of a micro grid using the OEM capacities can sometimes fail.::

        setting_batch_capacity      = True

In- and output files
____________________
Defining, whether or not input files are to be used for weather (not jet functional) and demand with **use_input_file_demand** and **use_input_file_weather**. If enabled, the file path should be specified with _input_file_demand_ and _input_file_weather_. In input_values.py unit and path of the data in the .csv file have to be specified (_unit_of_input_file_ Wh/kWh) to normalized the values in regards to the energy system, which performs calculations based on kWh.::

        use_input_file_demand       = True
        use_input_file_weather      = False

If no input files are used for demand and weather, the calculated demand and irradiation series can be saved by enabling **write_demand_to_file**, **write_demand_to_file**. The output folder and file prefix (**output_folder** and **output_file**) is defined further below. Notice, that all oemof simulation results are saved in the output folder. The files can be quite numerous, if a sensitivity analysis is performed, but each file is named explicitly after the sensitivity parameters used to generate the results.::

        write_demand_to_file = False
        write_weather_to_file = False
        if use_input_file_demand == True: write_demand_to_file = False
        if use_input_file_weather == True: write_weather_to_file = False

        output_folder='./simulation_results'
        output_file='results'

Display of results and graphs
______________________________
Oemof can generate and save .lp files and .oemof files with the simulation results. These can be saved to the output folder,if **setting_save_lp_file**, **setting_save_oemofresults** are set to True. Especially with long computing times, the oemof results should be saved.::

        setting_save_lp_file        = False
        setting_save_oemofresults   = True

During simulation, a number of graphs can be generated (**display_graphs_solar**, **display_graphs_demand**, **display_graphs_simulation**, **display interrupts computation**) and simulation results printed in the command line (**print_simulation_meta**, **print_simulation_main**, **print_simulation_invest**). If the performed simulation includes a sensitivity analysis, it is advisable to disable all these functions. The details of the simulated case can be displayed with **print_simulation_experiment**.::

        display_graphs_solar        = False
        display_graphs_demand       = False
        display_graphs_simulation   = True
        print_simulation_meta       = False
        print_simulation_main       = False
        print_simulation_invest     = False
        print_simulation_experiment = False

Results saved to csv-file::

        results_demand_characteristics      = True
        results_blackout_characteristics    = True
        results_annuities                   = True
        results_costs                       = True

Oemof settings
______________
In general, the solver of oemof is set to cbc (**solver**). The solver output (**solver_verbose**) is not shown if False.::

        solver = 'cbc'
        solver_verbose = False

To increase computation speed (especially for nonconvex flows in the dispatch OEM), an additional solver option **cmdline_option** is added  with value _cmdline_option_value_. It influences, when the solver accepts the found solution as optimal. Possible options:

* ratioGap
* allowedGap

Comand line options::

        cmdline_option       = 'ratioGap'
        cmdline_option_value = 0.01

The lp file of the energy system analysis with oemof can be saved as well (**setting_lp_file**). When debugging, one should set **debug** to True and limit the analysed timesteps (**coding_process** or even less (ie. 3) timesteps).::

        debug = True


Evaluated timeframe
--------------------
The results (NPV, LCOE, Annuity, fuel consumption) are scaled to **represent the real (annual) values** and are **not** the costs / consumption for the evaluated time period alone (see oemof_general.py)!::


        if coding_process == True:
            evaluated_days          =  1
            time_start              = pd.to_datetime('2018-07-07 0:00', format='%Y-%m-%d %H:%M')
            time_end       = time_start + pd.DateOffset(days=evaluated_days) - pd.DateOffset(hours=1)
            time_frequency          = 'H'

        else:
            evaluated_days          =  1
            time_start              = pd.to_datetime('2018-01-01 0:00', format='%Y-%m-%d %H:%M')
            time_end                = time_start + pd.DateOffset(days=evaluated_days) - pd.DateOffset(hours=1)
            time_frequency          = 'H'

        date_time_index = pd.date_range(start=time_start, end=time_end, freq=time_frequency)

Currently, the tool can ONLY evaluate hourly timesteps.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is inspired from [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and the versioning aim to respect [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

Here is a template for new release sections

```
## [_._._] - 20XX-MM-DD

### Added
-
### Changed
-
### Removed
-
### Fixed
-
```

## [Unreleased]

### Added
-
### Changed
-
### Removed
-
### Fixed
-

## [Offgridders V4.6] - 2020-08-17

### Added
-Docstrings to A1,B (#131,#132)


### Changed
- Replaced `== False` with `is False` and `== True` with `is True`
- Facilitated upgrade to unbundled oemof version: `oemof.solph == 0.4.0`, `oemof.network==0.4.0rc0`, `oemof.tools==0.4.0` and update `requirements.txt`
- Update README (#129)

### Added
- Literature file to readthedocs (#129)
- Added some docstrings to functions in `G2a`, copied from `Components.rst` (#125)
- Replaced "Yes" and "No" with True and False. In H0 and inside the Excel Files (#127)

### Fixed
-Small content corrections to readthedocs (#119,#123,#125)


## [Offgridders V4.5] - 2020-08-12

### Added
- Constant variables for strings are gathered in `constants.py` (#89,#117, #118)

### Changed
- Replaced parameter strings with variable calls  (#89,#117, #118)
- Changed path calls to `os.path.join` to enable compatability (#118)
- Log messages now call variable of parameter name (#117, #118)

### Removed
- Try/Except clauses at module import (#103) 

### Fixed
-

## [Offgridders V4.4] - 2020-07-09

### Added
- Error message if parameter `evaluation_perspective` is not chosen correctly (#97)
- Warning message if `fuel_price_change_annual` != 0, as calculation may be faulty (#97)

### Changed
- Added parameter `consumption_fuel_annual_kWh` to simulation outputs (#97)
- Input template file so that it runs MCA and gives options of `evaluation_perspective` (#97)
- If `fuel_price_change_annual` == 0, function `present_value_changing_fuel_price` is not executed and returns fuel price (#97)
- Order of cost results in simulation outputs, as `first_investment_cost`, `operation_mantainance_expenditures` as these are not to be added to costs, expenditures and revenues to calculate the NPC (#97)
- Formula calculating the residual value of an asset. Now, the sales revenue is translated into a present value: 
`linear_depreciation_last_investment = last_investment / lifetime` and `capex = capex - linear_depreciation_last_investment * (number_of_investments * lifetime - project_life) / (1 + wacc) ** (project_life)` (#97)
                )`
### Removed
- Removed unused function call in G2a (`genset_oem_minload`) (#97)

### Fixed
- Miscalculation of `total_demand_supplied_annual_kWh` due to wrong `evaluation_perspective` on tab `case_definitions`  (#97)
- Function call of `present_value_changing_fuel_price` (#97)
- Typo: Replace `operation_mantainance_expenditures` by `operation_maintenance_expenditures` (#97)

## [Offgridders V4.3] - 2020-07-02

### Fixed
- Faulty function calls from module structure change: 'get_universal_parameters',
'get_combinations_around_base', 'get_number_of_blackouts' (#86)

## [Offgridders V4.2] - 2020-06-30

### Added
- `total_excess_annual_kWh` to list of output parameters. Calculation: `total_excess_annual_kWh = total_excess_ac_annual_kWh + total_excess_dc_annual` (#73)
- Optional parameters and their default values `fuel_co2_emission_factor` (2.68 kgCO2eq/l diesel) and `maingrid_co2_emission_factor` (0.9 kgCO2eq/kWh) to input template, sheet `input_constant` and module `B` (#77)
- Calculation of C02 emissions and new output value: `co2_emissions_kgCO2eq`. It is based on the kWh consumption from the national grid (before transfromer station losses) and diesel consumption (#77)

### Changed
-

### Removed
-

### Fixed
- Typo when calling for the inverter capacity ('capacity_inverter_dc_ac_kW' and not 'capacity_inverter_kW'), (#75)
- Typo for unit of `maingrid_extension_lifetime` (a)

## [Offgridders V4.1] - 2020-06-30

### Added
- readthedocs.yml file (#59)

### Changed
- Moved Wiki to Readthedocs (#41)
- Changed class structure to modules/functions (#53)
- Script A0 was refactored to Offgridders. It runs now through `python Offgridders.py YOUR_INPUT_EXCEL_SHEET_PATH` (#53)

### Fixed
- Compilation of readthedocs by changing advanced settings on readthedocs.io (#59)

## [Offgridders V4.0] 2020-04-30

### Added
- Multi-Criteria Analysis (MCA), a post-processing comparison of the optimization results with H0 and H1
- Plotting bar charts of most relevant criteria
- Calculation of total first time investment costs ("first_investment") and total O&M costs ("operation_mantainance_expenditures")

### Changed
- Changelog.md was updated with old release infos
- requirements.txt: added scipy, xlsxwriter 
- test_input_template.xlsx: Setting whether to perform the multicriteria analysis (MCA) in settings tab; Electricity sales tariff (for MCA only) in case_definitions
- test_input_template.xlsx: New tab "multicriteria_data": weights of dimensions and criteria (all with default values) and which sensibility parameters should be included in the MCA
- A0 to call MCA
- B to read data from xlsx for MCA
- C: Extend set of parameters that are an output of simulation by "total_pv_generation_kWh", "total_wind_generation_kWh", "total_genset_generation_kWh"

### Removed

## Offgridders V3.1.1 [2020-04-01]

Intermediary release, maily focussing on restructuring the code into a code folder and adding a readthedocs.
Readthedocs is still dysfunctional. When starting to use Offgridders, please use release v3.1.

## Offgridders v3.1 [2020-04-01]
Release that includes all changes up to October 2019
- Version that is used for PSBs Cobrador paper + Book chapter
- Version that is used for MMHs Backup system paper

Info: Up to this point, no Changelog.md was in place. Changes will be tracked more carefully starting v4.0

## oesmot v3.0 (22.07.2019)
Working title of tool:

config.py
import pandas as pd

####### ------------ General simulation settings ----------------- #######
coding_process                  = True  # Defines timeframe and noise (see below)
restore_oemof_if_existant       = False  # If set to False, the directory with results is emptied!!
restore_blackouts_if_existant   = False
base_case_with_min_loading      = False   # If set to True, the generator capacity will be equal to peak demand in kW!

# TOdo allow multiple generators here (or even allow multiple generators in general??)

# simulated cases:
simulated_cases = {
    'offgrid_fixed': True,             # dispatch with base oem capacities
    'interconnected_buy': True,        # dispatch with interconnected main grid (only consumption)
    'interconnected_buysell': True,    # dispatch with interconnected main grid (consumption+feedin)
    'oem_grid_tied_mg': True,          # optimization of grid-tied mg (consumption+feedin)
    'sole_maingrid': True,             # supply only by main grid (allowed shortage variable)
    'buyoff': False, # todo not implemented
    'parallel': False, # todo not implemented
    # todo this could also be 1year without grid interconnection and one year with interconnection...?
    'adapted': False, # todo not implemented
}

# oemof simulation
allow_shortage               = False  # Allow supply shortage, details given below
setting_batch_capacity       = True
include_stability_constraint = True
setting_pcc_utility_owned    = True # if true, then consumption/feed in is counted from mg side, ie. transformer loss is revenue loss of utility

####### ----------------- Settings for files------------------- #######
# Input (files)
use_input_file_demand       = True
use_input_file_weather      = True
# Outputs
write_demand_to_file = False
write_weather_to_file = False
if use_input_file_demand == True: write_demand_to_file = False
if use_input_file_weather == True: write_weather_to_file = False

# Output file paths
output_folder='./simulation_results'
output_file='results'

# display results and graphs
setting_save_lp_file                = True  # save lp file of oemof simulation
setting_save_oemofresults           = True   # save oemofresults to .oemof file
display_graphs_solar                = False
display_graphs_demand               = False
display_graphs_flows_storage        = True
display_graphs_flows_electricity_mg = True

print_simulation_meta       = False  # print information on opimization
print_simulation_main       = False  # print accumulated flows over electricity bus
print_simulation_invest     = False  # print investment results
print_simulation_experiment = False  # Print data on experiment run (sensitivity analysis)

# Choose flows saved to csv
setting_save_flows_storage          = True
setting_save_flows_electricity_mg   = True

# Choose results saves in csv
results_demand_characteristics      = True
results_blackout_characteristics    = True
results_annuities                   = True
results_costs                       = True

####### ----------------- Oemof simulation settings ------------------- #######
# Define solver
solver = 'cbc'
solver_verbose       = False  # show/hide solver output
cmdline_option       = 'ratioGap' #  options for solver: allowedGap,  mipgap, ratioGap
cmdline_option_value = 1*10**(-1)

# Debugging

####### ----------------- Input data for settings  ------------------- #######

# Simulation timeframe
if coding_process == True:
    evaluated_days  =  1
    time_start      = pd.to_datetime('2018-07-07 00:00', format='%Y-%m-%d %H:%M')
    time_end        = time_start + pd.DateOffset(days=evaluated_days) - pd.DateOffset(hours=1)
    #time_end = pd.to_datetime('2018-07-07 15:00', format='%Y-%m-%d %H:%M')
    time_frequency  = 'H'
    debug = True  # Set number_of_timesteps to 3 to get a readable lp-file.


else:
    evaluated_days = 365
    time_start     = pd.to_datetime('2018-01-01 0:00', format='%Y-%m-%d %H:%M')
    time_end       = time_start + pd.DateOffset(days=evaluated_days) - pd.DateOffset(hours=1)
    time_frequency = 'H'
    # date_time_index.freq checkts for frequency, but result is sth like <Hour> or <15 * Minutes>
    debug = False  # Set number_of_timesteps to 3 to get a readable lp-file.

date_time_index = pd.date_range(start=time_start, end=time_end, freq=time_frequency)